### PR TITLE
Add raytracing to the egg finder to make it fairer

### LINKED ISSUE #832: Chocolate Eggs have ESP
### What happened?

The chocolate egg finder detects armor stands through walls impossible to see by the player. Hypixel does not really allow ESP like that, a proper approach would be to use the particles from the egglocater instead.

### Screenshots

image
This egg was detected without me ever seeing it or getting it shared in chat.

### Log output

https://mclo.gs/J1NFRDp

### Minecraft Version

1.21.0

### Skyblocker Version

1.21.1+1.21

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Emirlol**
I'm honestly not sure how to explain the changes. Might be better to read commits.
Fixes #832 by adding ray tracing every tick. Also has better logic for waypoints and chat messages, now it doesn't highlight eggs/send messages when the egg was already found in another island, and this state is reset when the egg's reset hour happens, which is at 7, 14, 21 (7am, 2pm, 9 pm)

Also cleans up and improves `SkyblockTime`. Now it includes hour and refreshes every 50 secs (a skyblock hour). There are also events for when the time unit changes value.

### CODE REVIEW

**Comment by Vricken** [comment]
Could the raycast be an option? Like a setting called “enable egg waypoints without seeing them (possibly bannable use at your own risk)” or something like that

**Comment by Julienraptor01** [comment]
> Could the raycast be an option? Like a setting called “enable egg waypoints without seeing them (possibly bannable use at your own risk)” or something like that

no, it's not ESP is a cheat and it's an accident that we included such behavior
tho it's fully open source, so feel free to unpatch it, but it won't be released as an option
(this is just my opinion but i think no one on the team want the mod to become a cheat)

**Comment by Emirlol** [comment]
> > Could the raycast be an option? Like a setting called “enable egg waypoints without seeing them (possibly bannable use at your own risk)” or something like that
> 
> no, it's not, ESP is a cheat and it's an accident that we included such behavior tho it's fully open source, so feel free to unpatch it, but it won't be released as an option (this is just my opinion but i think no one on the team want the mod to become a cheat)

Exactly this. I just never had the thought of whether it was cheaty or not when making it. Took a while to fix due to some skill issues and laziness. This PR is how it should've been since the start.

I know it's currently a bit hard to use, as all it does is tell you if you unobstructed vision to eggs. I was also planning on adding labels for possible locations and egg locator helper for that, but I have other stuff to do for now so those will come separately later on.

File: src/main/java/de/hysky/skyblocker/skyblock/chocolatefactory/EggFinder.java [hunk 1]
```diff
@@ -39,18 +40,39 @@ public class EggFinder {
...
 
 	public static void init() {
-		ClientPlayConnectionEvents.JOIN.register((ignored, ignored2, ignored3) -> invalidateState());
+		ClientPlayConnectionEvents.JOIN.register((ignored, ignored2, ignored3) -> isLocationCorrect = false);
 		SkyblockEvents.LOCATION_CHANGE.register(EggFinder::handleLocationChange);
 		ClientReceiveMessageEvents.GAME.register(EggFinder::onChatMessage);
 		WorldRenderEvents.AFTER_TRANSLUCENT.register(EggFinder::renderWaypoints);
+		ClientTickEvents.END_CLIENT_TICK.register(client -> {
+			if (!SkyblockerConfigManager.get().helpers.chocolateFactory.enableEggFinder || client.player == null) return;
+			for (EggType type : EggType.entries) {
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/chocolatefactory/EggFinder.java:64]: there should be some more checks to ensure that this isn't constantly creating an iterator, such as being in skyblock, correct location & time

**Review by AzureAaron — CHANGES_REQUESTED** [review]

File: src/main/java/de/hysky/skyblocker/utils/SkyblockTime.java [hunk 2]
```diff
@@ -58,4 +99,64 @@ public enum Month {
...
+		for (OnYearChange listener : listeners) {
+			listener.onYearChange(year);
+		}
+	});
+
+	public static final Event<OnTimeUpdate> TIME_UPDATE = EventFactory.createArrayBacked(OnTimeUpdate.class, listeners -> (year, season, month, day, hour) -> {
+		for (OnTimeUpdate listener : listeners) {
+			listener.onTimeUpdate(year, season, month, day, hour);
+		}
+	});
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/utils/SkyblockTime.java:161]: These should go at the top. Maybe in a block after the logger?

**Review by kevinthegreat1 — CHANGES_REQUESTED** [review]

**Review by kevinthegreat1 — APPROVED** [review]

**Review by AzureAaron — APPROVED** [review]

**Pull request merged**